airflow/migrations/versions/0143_2_9_2_add_indexes_on_dag_id_column_in_referencing_tables.py

```diff
@@ -0,0 +1,115 @@
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+"""Add indexes on dag_id column in referencing tables.
+
+Revision ID: 0fd0c178cbe8
+Revises: 686269002441
+Create Date: 2024-05-15 16:52:39.077349
+
+"""
+
+from __future__ import annotations
+
+from alembic import op
+
+# revision identifiers, used by Alembic.
+revision = "0fd0c178cbe8"
+down_revision = "686269002441"
+branch_labels = None
+depends_on = None
+airflow_version = "2.9.2"
+
+
+def upgrade():
+    """Apply Add indexes on dag_id column in referencing tables."""
+    with op.batch_alter_table("dag_owner_attributes", schema=None) as batch_op:
+        batch_op.create_index("idx_dag_owner_attributes_dag_id", ["dag_id"], unique=False)
+
+    with op.batch_alter_table("dag_schedule_dataset_reference", schema=None) as batch_op:
+        batch_op.create_index("idx_dag_schedule_dataset_reference_dag_id", ["dag_id"], unique=False)
+
+    with op.batch_alter_table("dag_tag", schema=None) as batch_op:
+        batch_op.create_index("idx_dag_tag_dag_id", ["dag_id"], unique=False)
+
+    with op.batch_alter_table("dag_warning", schema=None) as batch_op:
+        batch_op.create_index("idx_dag_warning_dag_id", ["dag_id"], unique=False)
+
+    with op.batch_alter_table("dataset_dag_run_queue", schema=None) as batch_op:
+        batch_op.create_index("idx_dataset_dag_run_queue_target_dag_id", ["target_dag_id"], unique=False)
+
+    with op.batch_alter_table("task_outlet_dataset_reference", schema=None) as batch_op:
+        batch_op.create_index("idx_task_outlet_dataset_reference_dag_id", ["dag_id"], unique=False)
+
+
+def _handle_foreign_key_constraint_index_deletion(
+    batch_op, constraint_name, index_name, local_fk_column_name
+):
+    batch_op.drop_constraint(constraint_name, type_="foreignkey")
+    batch_op.drop_index(index_name)
+    batch_op.create_foreign_key(
+        constraint_name, "dag", [local_fk_column_name], ["dag_id"], ondelete="CASCADE"
+    )
+
+
+def downgrade():
+    """Unapply Add indexes on dag_id column in referencing tables."""
+    # conn = op.get_bind()
+    # with op.batch_alter_table("dag_owner_attributes", schema=None) as batch_op:
+    #     if conn.dialect.name == "mysql":
+    #         batch_op.execute("ALTER TABLE dag_owner_attributes DROP FOREIGN KEY `dag.dag_id`;")
+    #         batch_op.drop_index("idx_dag_owner_attributes_dag_id")
+    #         batch_op.create_foreign_key("dag.dag_id", "dag", ["dag_id"], ["dag_id"], ondelete="CASCADE")
+    #     else:
+    #         _handle_foreign_key_constraint_index_deletion(
+    #             batch_op, "dag.dag_id", "idx_dag_owner_attributes_dag_id", "dag_id"
+    #         )
```

Comment:
I have commented this block to check if everything else runs fine in the CI.
This block when uncommented is failing in CI however the same runs fine locally with breeze on all database backends MySQL, PostgresSQL and SQLite. It fails saying that it cannot find the constraint `dag.dag_id` for dropping. However, when I check locally in the db, that table does have that constraint. 
The diff for this is that when the ForeignKey is declared in the model, it is defined with name `dag.dag_id` and I'm guessing this is somehow affecting finding the constraint. I will investigate this further.

Reply:
From the earlier failure, dropping foreignkeys should only be done for MySQL, right? I see that we drop it for others here as well but should be for MySQL only

Reply:
yes, however this does not harm us IMO. We can however restrict that specifically for MySQL as an improvisation. 

But the failure I mentioned [here](https://github.com/apache/airflow/pull/39638#discussion_r1602152626) still needs to be investigated

Reply:
The problem is in your DDL, in postgres, there's no DROP FOREIGN KEY. It's a constraint. So you have to do `alter table tablename drop constraint constraint_name` better still you can use alembic's drop_contraint for all DB backends. `batch_op.drop_constraint('name', type_="foreignkey")

Reply:
Please let's only drop foreign keys for MYSQL. Others seems fine without dropping of foreign keys

Reply:
> The problem is in your DDL, in postgres, there's no DROP FOREIGN KEY. It's a constraint. So you have to do alter table tablename drop constraint constraint_name better still you can use alembic's drop_contraint for all DB backends. batch_op.drop_constraint('name', type_="foreignkey")

@ephraimbuddy no, that's not my DDL for postgres. And that's not even the DDL even for mysql for the other 5 tables below.

If you review the PR files, we have the common code method that we're calling which in-fact uses alembic's drop_constraint

```
def _handle_foreign_key_constraint_index_deletion(
    batch_op, constraint_name, index_name, local_fk_column_name
):
    batch_op.drop_constraint(constraint_name, type_="foreignkey")
    batch_op.drop_index(index_name)
    batch_op.create_foreign_key(
        constraint_name, "dag", [local_fk_column_name], ["dag_id"], ondelete="CASCADE"
    )
```

I added the DROP FOREIGN KEY for experimentation but only for mysql and only for one specific table `dag_owner_attributes`. On the same table `dag_owner_attributes`, we still call `_handle_foreign_key_constraint_index_deletion` for PostgreSQL and SQLite which leverages alembic's drop_constraint. 

The problem like I mentioned in my first [comment](https://github.com/apache/airflow/pull/39638#discussion_r1602152626) is that the DB backends in our CI are not able to find the constraint `dag.dag_id` for the `dag_owner_attributes` table. The CI is happy for all the other 5 tables.

Reply:
I see what you are saying. How about you don't add the quotes on the "dag.dag_id"? Not saying it would work but maybe it's seeing that differently

Reply:
I am not adding additional quotes as such. It's that drop_constraint needs the name of the constraint which it takes as a string argument and hence to make it a string need to pass in quotes similar to other constraint names we pass. While creating the foreign key as well, we specify the same constraint name "dag.dag_id" as a string https://github.com/apache/airflow/blob/cba653ee23c1e596855cdc620fe893fecb3a6a44/airflow/models/dag.py#L3572

Reply:
Well, you can create an issue for it but let's work this one to fix MySQL and not drop foreignkey for postgres/sqlite except if they don't work

Reply:
yes, makes sense. I will follow-up for the index on `dag_owner_attributes` table in a separate PR. Also updated the PR description to mention the same. I have addresed the other feedback to drop & recreate foreign key constraints only in the case of MySQL. please check.